feat: add `math/base/special/wrapf`

gururaj1512 opened the pull request:
Resolves a part of #649

## Description

> What is the purpose of this pull request?
- adds `math/base/special/wrapf`, which would be the single-precision equivalent for [math/base/special/wrap](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/bernoulli).

```c
float stdlib_base_wrapf( const float v, const float min, const float max );
```

## Related Issues

> Does this pull request have any related issues?

This pull request:

- Progresses: #649 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

gururaj1512 committed:
feat: add wrapf

gururaj1512 commented:
Hi @gunjjoshi, could you review this PR? Some checks are failing, need guidance for it.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/src/main.c:
```diff
@@ -0,0 +1,70 @@
+	if ( vc < minc ) {
+		vc += delta * ( stdlib_base_truncf( ( minc - vc ) / delta ) + 1.0 );
+	}
+	return minc + ( fmodf( vc - minc, delta ) );
```
@gururaj1512 The `fmodf` which is being used here, is not being included at the top. We are currently missing the implementation of `fmodf`. You'll need to include it once we have added `fmodf`.

gururaj1512 replied:
@gunjjoshi, Thanks for clarifying

Planeshifter replied:
@gururaj1512 `fmodf` (added by @gunjjoshi in https://github.com/stdlib-js/stdlib/pull/3059) has now landed, so this branch can be rebased on the latest code on `develop` and updated to make use of it.

gururaj1512 committed:
Merge branch 'stdlib-js:develop' into wrapf

gururaj1512 committed:
update main.c

gururaj1512 committed:
update manifest.json

gururaj1512 committed:
update manifest.json

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into wrapf

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/README.md:
```diff
@@ -0,0 +1,255 @@
+
+#### wrapf( v, min, max )
+
+Wraps a value on the half-open interval `[min,max)`.
```
This must be updated to indicate single-precision floating-point numbers, same as the description above.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/README.md:
```diff
@@ -0,0 +1,255 @@
+int main( void ) {
+    const float min[] = { 0.0, 0.0, 0.0, 0.0, -3.14 };
+    const float max[] = { 5.0, 5.0, 5.0, 5.0, -0.0 };
+    const float v[] = { 3.14, -3.14, 10.0, -0.0, 0.0 };
```
We must add a `f` suffix after each of these numbers, as they are single-precision floating-point numbers.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/README.md:
```diff
@@ -0,0 +1,255 @@
+
+-   <span class="package-name">[`@stdlib/math/base/special/clampf`][@stdlib/math/base/special/clampf]</span><span class="delimiter">: </span><span class="description">restrict a single-precision floating-point number to a specified range.</span>
+
+</section>
```
This section should be kept empty, as it is populated automatically.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/README.md:
```diff
@@ -0,0 +1,255 @@
+
+[@stdlib/math/base/special/clampf]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/clampf
+
+<!-- </related-links> -->
```
Same goes for this one as well.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/benchmark/benchmark.js:
```diff
@@ -0,0 +1,51 @@
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		x = ( randu()*20.0 ) - 10.0;
+		y = wrapf( x, -5.0, 5.0 );
```
We should use `random/array/uniform` here, and generate the array outside the loop. See: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/minf/benchmark/benchmark.js.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/benchmark/benchmark.native.js:
```diff
@@ -0,0 +1,60 @@
+	b.tic();
+	for ( i = 0; i < b.iterations; i++ ) {
+		x = ( randu()*20.0 ) - 10.0;
+		y = wrapf( x, -5.0, 5.0 );
```
Same comment here as well.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,153 @@
+	// Perform range reduction...
+	delta = max - min;
+	if ( v < min ) {
+		v += delta * ( truncf( (min-v)/delta ) + 1.0 );
```
```suggestion
		v += delta * ( truncf( ( min - v ) / delta ) + 1.0 );
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,153 @@
+	if ( v < min ) {
+		v += delta * ( truncf( (min-v)/delta ) + 1.0 );
+	}
+	return min + fmodf( (v-min), delta );
```
```suggestion
	return min + fmodf( ( v - min ), delta );
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,153 @@
+*
+* @return elapsed time in seconds
+*/
+static double benchmark( void ) {
+	double elapsed;
+	double t;
+	float x;
+	float y;
+	int i;
+
+	t = tic();
+	for ( i = 0; i < ITERATIONS; i++ ) {
+		x = ( 20.0*rand_float() ) - 10.0;
+		y = wrapf( x, -5.0, 5.0 );
```
Similar to the comment on JS benchmarks, here, too, we can declare `x` as an array, and initialize values outside this loop. Refer: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/minf/benchmark/c/native/benchmark.c#L92

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,133 @@
+static double benchmark( void ) {
+	double elapsed;
+	double t;
+	float x;
+	float y;
+	int i;
+
+	t = tic();
+	for ( i = 0; i < ITERATIONS; i++ ) {
+		x = ( 20.0*rand_float() ) - 10.0;
+		y = stdlib_base_wrapf( x, -5.0, 5.0 );
```
Same comment here as well.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/benchmark/c/Makefile:
We can use the newer Makefile here.
See: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/minf/benchmark/c/native/Makefile

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/examples/c/Makefile:
```diff
@@ -0,0 +1,146 @@
+#/
+# @license Apache-2.0
+#
+# Copyright (c) 2022 The Stdlib Authors.
```
```suggestion
# Copyright (c) 2024 The Stdlib Authors.
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/examples/c/example.c:
```diff
@@ -0,0 +1,33 @@
+#include <stdio.h>
+
+int main( void ) {
+	const float min[] = { 0.0, 0.0, 0.0, 0.0, -3.14 };
+	const float max[] = { 5.0, 5.0, 5.0, 5.0, -0.0 };
+	const float v[] = { 3.14, -3.14, 10.0, -0.0, 0.0 };
```
Missing `f` suffixes.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/lib/main.js:
We must use `float64ToFloat32` here, and use it to wrap the inputs and arithmetic operations.
For reference, see: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/logf/lib/main.js

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/lib/main.js:
```diff
@@ -0,0 +1,100 @@
+		max <= min
+	) {
+		return NaN;
+	}
+	// Normalize +-0 to +0...
```
```suggestion
	}

	// Normalize +-0 to +0...
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/lib/main.js:
```diff
@@ -0,0 +1,100 @@
+	}
+	if ( max === 0.0 ) {
+		max = 0.0;
+	}
+	// Simple case where value is already within range...
```
```suggestion
	}

	// Simple case where value is already within range...
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/lib/main.js:
```diff
@@ -0,0 +1,100 @@
+	// Simple case where value is already within range...
+	if ( min <= v && v < max ) {
+		return v;
+	}
+	// Perform range reduction...
```
```suggestion
	}

	// Perform range reduction...
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/lib/main.js:
```diff
@@ -0,0 +1,100 @@
+	// Perform range reduction...
+	delta = max - min;
+	if ( v < min ) {
+		v += delta * ( truncf( (min-v)/delta ) + 1.0 );
+	}
+	return min + ( (v-min) % delta );
```
The spacing is off here.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/lib/native.js:
```diff
@@ -0,0 +1,72 @@
+*
+* @example
+* var v = wrapf( 3.14, 0.0, 5.0 );
+* // returns 3.14
```
```suggestion
* // returns ~3.14
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/src/Makefile:
```diff
@@ -0,0 +1,70 @@
+#/
+# @license Apache-2.0
+#
+# Copyright (c) 2022 The Stdlib Authors.
```
```suggestion
# Copyright (c) 2024 The Stdlib Authors.
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/src/main.c:
```diff
@@ -0,0 +1,71 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2022 The Stdlib Authors.
```
```suggestion
* Copyright (c) 2024 The Stdlib Authors.
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/src/main.c:
```diff
@@ -0,0 +1,71 @@
+* @return     wrapped value
+*
+* @example
+* float v = stdlib_base_wrapf( 3.14, 0.0, 5.0 );
```
Missing `f` suffixes here and elsewhere in this file.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/src/main.c:
```diff
@@ -0,0 +1,71 @@
+	}
+	if ( stdlib_base_is_negative_zerof( maxc ) ) {
+		maxc = 0.0;
+	}
+	// Simple case where value is already within range...
```
```suggestion
	}

	// Simple case where value is already within range...
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/src/main.c:
```diff
@@ -0,0 +1,71 @@
+	// Simple case where value is already within range...
+	if ( minc <= vc && vc < maxc ) {
+		return vc;
+	}
+	// Perform range reduction...
```
```suggestion
	}

	// Perform range reduction...
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/test/test.js:
```diff
@@ -0,0 +1,167 @@
+tape( 'if provided `NaN` for any argument, the function returns `NaN`', function test( t ) {
+	var v;
+
+	v = wrapf( NaN, 0.0, 5.0 );
+	t.strictEqual( isnanf( v ), true, 'returns NaN' );
+
+	v = wrapf( 0.0, NaN, 5.0 );
+	t.strictEqual( isnanf( v ), true, 'returns NaN' );
+
+	v = wrapf( 3.14, 0.0, NaN );
+	t.strictEqual( isnanf( v ), true, 'returns NaN' );
+
+	v = wrapf( NaN, NaN, 5.0 );
+	t.strictEqual( isnanf( v ), true, 'returns NaN' );
+
+	v = wrapf( NaN, 0.0, NaN );
+	t.strictEqual( isnanf( v ), true, 'returns NaN' );
+
+	v = wrapf( 3.14, NaN, NaN );
+	t.strictEqual( isnanf( v ), true, 'returns NaN' );
+
+	v = wrapf( NaN, NaN, NaN );
+	t.strictEqual( isnanf( v ), true, 'returns NaN' );
```
We should replace:
```
returns NaN
```
by:
```
returns expected value
```
here.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/wrapf/test/test.native.js:
```diff
@@ -0,0 +1,176 @@
+tape( 'if provided `NaN` for any argument, the function returns `NaN`', opts, function test( t ) {
+	var v;
+
+	v = wrapf( NaN, 0.0, 5.0 );
+	t.strictEqual( isnanf( v ), true, 'returns NaN' );
+
+	v = wrapf( 0.0, NaN, 5.0 );
+	t.strictEqual( isnanf( v ), true, 'returns NaN' );
+
+	v = wrapf( 3.14, 0.0, NaN );
+	t.strictEqual( isnanf( v ), true, 'returns NaN' );
+
+	v = wrapf( NaN, NaN, 5.0 );
+	t.strictEqual( isnanf( v ), true, 'returns NaN' );
+
+	v = wrapf( NaN, 0.0, NaN );
+	t.strictEqual( isnanf( v ), true, 'returns NaN' );
+
+	v = wrapf( 3.14, NaN, NaN );
+	t.strictEqual( isnanf( v ), true, 'returns NaN' );
+
+	v = wrapf( NaN, NaN, NaN );
+	t.strictEqual( isnanf( v ), true, 'returns NaN' );
```
Same comment here too.

gururaj1512 committed:
refactor: as per suggested

gururaj1512 committed:
fix: benchmark and test file

gururaj1512 committed:
fix: lint errors

gururaj1512 committed:
fix: errors

gunjjoshi commented:
@gururaj1512 Can you go through the CI failures and make necessary changes in this PR? There are some linting errors and test failures.

gururaj1512 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into wrapf

gururaj1512 committed:
fix: lint errors

gururaj1512 commented:
@gunjjoshi can you review the implementaion, I am unable to point out the cause of unsuccessful test cases

gururaj1512 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into wrapf

gururaj1512 committed:
feat: fix test cases

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

gururaj1512 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

gururaj1512 committed:
fix: lint errors

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

anandkaranubc commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into wrapf

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/wrapf/lib/main.js:
```diff
@@ -0,0 +1,104 @@
+// MAIN //
+
+/**
+* Wraps a value on the half-open interval `[min,max)` for single-precision floating-point numbers.
```
```suggestion
* Wraps a single-precision floating-point value to the half-open interval `[min,max)`.
```

anandkaranubc committed:
refactor: apply suggestions from PR review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/math/base/special/wrapf/README.md:
```diff
@@ -0,0 +1,245 @@
+
+#### stdlib_base_wrapf( v, min, max )
+
+Wrap a single-precision floating-point value to the half-open interval `[min,max)`.
```
```suggestion
Wraps a single-precision floating-point value to the half-open interval `[min,max)`.
```

kgryte committed:
docs: fix description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/math/base/special/wrapf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,156 @@
+* @param max    maximum value
+* @return       wrapped value
+*/
+float wrapf( float v, float min, float max ) {
```
```suggestion
static float wrapf( float v, float min, float max ) {
```

kgryte commented on lib/node_modules/@stdlib/math/base/special/wrapf/benchmark/c/benchmark.c:
```diff
@@ -0,0 +1,156 @@
+	if ( v < min ) {
+		v += delta * ( truncf( (min-v)/delta ) + 1.0f );
+	}
+	return min + fmodf( (v-min), delta );
```
```suggestion
	return min + fmodf( v-min, delta );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/math/base/special/wrapf/include/stdlib/math/base/special/wrapf.h:
```diff
@@ -0,0 +1,38 @@
+}
+#endif
+
+#endif // !STDLIB_MATH_BASE_SPECIAL_WRAP_H
```
```suggestion
#endif // !STDLIB_MATH_BASE_SPECIAL_WRAPF_H
```

kgryte committed:
docs: fix comment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/math/base/special/wrapf/lib/main.js:
```diff
@@ -0,0 +1,111 @@
+	// Perform range reduction...
+	delta = f32( max - min );
+	if ( v < min ) {
+		v = f32( v + f32( delta * f32( truncf( f32(min-v) / f32(delta) ) + ONE ) ) );
```
```suggestion
		v = f32( v + f32( delta * f32( truncf( f32( f32(min-v) / delta ) ) + ONE ) ) );
```

kgryte committed:
refactor: move parens

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
